doc/tutorials/content/gpu_octree.rst

```diff
@@ -0,0 +1,235 @@
+
+
+.. note::
+
+    The GPU K nearest neighbor algorithm differs from it's CPU counterpart in implementation. Specifically, the CPU algorithm will only find approximate nearest neighbors, where the user can set the upper bound of the error allowed. In contrast, the GPU algorithm finds the absolute nearest neighbor.
+
+In order to perform K nearest search, we must prepare output buffers on the device to store the results of the search queries. Specifically, we need two separate buffers to store the resulting nearest neighbor indices and the corresponding squared distances between the search point and nearest neighbors. Note that these buffers will be automatically resized to `k * number of queries` size, in order to hold all the results.
+
+Now, we can perform the K nearest neighbors search on GPU. After performing the search we must download the results from GPU memory into host memory using two new vectors for indices and square distances.
+
+
+.. literalinclude:: sources/gpu_octree/gpu_octree.cpp
+    :language: cpp
+    :lines: 62-98
+
+Radius Search
+-------------
+
+Radius search returns neighbors within a given radius to the query point. The user must specify a maximum limit of points to return per query. There are implementations radius search available which either return / do not return the corresponding squared distances to the query point. Note that the methods which do not return squared distances are more efficient than those that do. In usage, this works very similar to the K nearest neighbor search.
+
+Neighbors are returned as a `pcl::gpu::NeighborIndices` object, which contains all result indices packed into a one dimensional vector called `data`, as well as an additional vector called `sizes` which contains the number of results found for each individual query. These sizes can can be used to split the resulting indices as well as square distances by query.
+
+Radius search can be conducted with a single radius for all queries or, individual radii may be specified for each query, by passing a `pcl::gpu::Octree::Radiuses` object of the same size as the queries object to the radius search method. In this tutorial we will specify individual radii for each of our ten queries.
+
+.. literalinclude:: sources/gpu_octree/gpu_octree.cpp
+    :language: cpp
+    :lines: 103-151
+
+
+Approximate Nearest Search
+--------------------------
+
+As the name suggests, approximate nearest search differs from K nearest search by the fact that it's traversal mechanism is 'approximate'. Specifically, during traversal, it will attempt to identify the most likely leaf to hold the nearest neighbor using the 'distance from query point to voxel center' heuristic. Thus there is no guarantee that this method will find the absolute nearest neighbor, and in fact there is also no guarantee of an upper-bound of the error in the result.
```

Comment:
It's important to benchmark runtime performance differences between the approximate method and the kNN? It might be another one of those cases in which it is in general faster.

Reply:
I ran the same query set for approxNearest and K nearest and actually the difference is huge (so much so that i ran it with two different data sizes to confirm)
```shell
test 1
points - 871000
queries - 100000
gpu-approx-nearest-batch-all took 0.265414ms
gpu-knn-batch-all took 111.781ms.

test 2 
points - 8710000
queries - 1000000
gpu-approx-nearest-batch-all took 2.0109ms.
gpu-knn-batch-all took 12039.1ms.
```
It looks like the approximation saves quite a lot of time (it makes sense because after the traversal only one single leaf has to be explored in the case of the approximate nearest search)

Reply:
This is very suspicious. A 4 order of magnitude difference it bounds already on the "there's a problem somewhere".

Reply:
Yeah it's possible. But I would argue that the number of points that the algorithm has to consider might also differ by the same magnitude (points in 1 leaf vs points in potentially 10s of 1000s of leafs, since voxels increase by powers of 8 at each level).